### PR TITLE
Implement presentation-time

### PR DESCRIPTION
**Pull request by emersion**
Test plan: `weston-presentation-shm` under DRM

Fixes https://github.com/swaywm/wlroots/issues/1172

### CODE REVIEW

File: types/meson.build [hunk 1]
```diff
@@ -19,13 +19,17 @@ lib_wlr_types = static_library(
 		'xdg_shell_v6/wlr_xdg_shell_v6.c',
 		'xdg_shell_v6/wlr_xdg_surface_v6.c',
 		'xdg_shell_v6/wlr_xdg_toplevel_v6.c',
+		'tablet_v2/wlr_tablet_v2_pad.c',
```
Comment (Ongy) [inline types/meson.build]: Is this alphabetical? X -> t -> w is weird, inside the subdirs it looks reasonable alphabetical.

File: types/wlr_output.c [hunk 2]
```diff
@@ -560,6 +560,17 @@ void wlr_output_schedule_frame(struct wlr_output *output) {
 		wl_event_loop_add_idle(ev, schedule_frame_handle_idle_timer, output);
 }
 
+void wlr_output_send_present(struct wlr_output *output, struct timespec *when,
+		unsigned seq, uint32_t flags) {
+	struct wlr_output_event_present event = {
+		.output = output,
```
Comment (Ongy) [inline types/wlr_output.c]: What's the reason for adding the output pointer here?
I'd expect it to be derivable from the actual handler that's called.

Comment (emersion) [reply types/wlr_output.c]: wlroots has no idea on which output the surface is on. How could you "derive it from the handler"?

Comment (emersion) [reply types/wlr_output.c]: Oh, yeah, sorry, I misunderstood. I just added it for consistency with other events. It's not strictly necessary I guess.

File: types/wlr_output.c [hunk 3]
```diff
@@ -18,6 +18,8 @@
 
 #define OUTPUT_VERSION 3
 
+#define DEFAULT_PRESENT_CLOCK CLOCK_MONOTONIC
```
Comment (SirCmpwn) [inline types/wlr_output.c]: -1

Comment (emersion) [reply types/wlr_output.c]: What do you mean?

Comment (SirCmpwn) [reply types/wlr_output.c]: This doesn't need to be abstracted, just use CLOCK_MONOTONIC explicitly.

Comment (emersion) [reply types/wlr_output.c]: There's more to come.

Comment (SirCmpwn) [reply types/wlr_output.c]: Okay, but I'm probably not going to change my mind.

**Comment by emersion** [comment]
This is ready for review.

File: backend/drm/drm.c [hunk 4]
```diff
@@ -1150,6 +1160,19 @@ static void page_flip_handler(int fd, unsigned seq,
 		post_drm_surface(&conn->crtc->primary->mgpu_surf);
 	}
 
+	struct timespec present_time = {
+		.tv_sec = tv_sec,
+		.tv_nsec = tv_usec * 1000,
+	};
+	struct wlr_output_event_present present_event = {
```
Comment (ammen99) [inline backend/drm/drm.c]: shouldn't `output` also be set?

Comment (emersion) [reply backend/drm/drm.c]: So this part is a little misleading. It's set in `wlr_output_send_present`. I just didn't want to create another struct. Maybe I can just set it here too, if that matters…

Comment (ammen99) [reply backend/drm/drm.c]: oh, yeah, I've missed that. Idk if something should be changed then ...

**Comment by emersion** [comment]
Rebased against master.

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**